d2/src/main/java/com/linkedin/d2/balancer/properties/ClusterStoreProperties.java

```diff
@@ -0,0 +1,191 @@
+        null, null);
+  }
+
+  public ClusterStoreProperties(String clusterName,
+      List<String> prioritizedSchemes,
+      Map<String, String> properties,
+      Set<URI> bannedUris,
+      PartitionProperties partitionProperties,
+      List<String> sslSessionValidationStrings,
+      Map<String, Object> darkClusters,
+      boolean delegated,
+      ClusterProperties canaryConfigs,
+      CanaryDistributionStrategy distributionStrategy)
+  {
+    super(clusterName, prioritizedSchemes, properties, bannedUris, partitionProperties, sslSessionValidationStrings, darkClusters, delegated);
+    _canaryConfigs = canaryConfigs;
+    _canaryDistributionStrategy = distributionStrategy;
+  }
+
+  public ClusterStoreProperties(ClusterProperties stableConfigs, ClusterProperties canaryConfigs, CanaryDistributionStrategy distributionStrategy)
+  {
+    super(stableConfigs.getClusterName(), stableConfigs.getPrioritizedSchemes(),
+        stableConfigs.getProperties(), stableConfigs.getBannedUris(), stableConfigs.getPartitionProperties(),
+        stableConfigs.getSslSessionValidationStrings(), stableConfigs.getDarkClusters(),
+        stableConfigs.isDelegated());
+    _canaryConfigs = canaryConfigs;
+    _canaryDistributionStrategy = distributionStrategy;
+  }
+
+  public ClusterProperties getCanaryConfigs()
+  {
+    return _canaryConfigs;
+  }
+
+  public CanaryDistributionStrategy getCanaryDistributionStrategy()
+  {
+    return _canaryDistributionStrategy;
+  }
+
+  public boolean hasCanary() {
+    return _canaryConfigs != null && _canaryDistributionStrategy != null;
+  }
+
+  /**
+   * Given a canary distribution (stable or canary), return the corresponding distributed/picked cluster properties.
+   */
+  public ClusterProperties getDistributedClusterProperties(CanaryDistributionProvider.Distribution distribution)
+  {
+    if (distribution.equals(CanaryDistributionProvider.Distribution.CANARY) && hasCanary())
+    {
+      return _canaryConfigs;
+    }
+    return new ClusterProperties(this); // make a copy of stable configs with the super class copy constructor
+  }
+
+  @Override
+  public String toString()
+  {
+    return "ClusterStoreProperties [_stableClusterProperties=" + super.toString() + ", _canaryConfigs=" + _canaryConfigs
+        + ", _canaryDistributionStrategy=" + _canaryDistributionStrategy + "]";
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int prime = 31;
+    int result = super.hashCode();
+    result = prime * result + ((_canaryConfigs == null) ? 0 : _canaryConfigs.hashCode());
+    result = prime * result + ((_canaryDistributionStrategy == null) ? 0 : _canaryDistributionStrategy.hashCode());
+    return result;
+  }
+
+  @Override
+  public boolean equals(Object obj)
+  {
+    if (!super.equals(obj))
+    {
+      return false;
+    }
+    ClusterStoreProperties other = (ClusterStoreProperties) obj;
```

Comment:
This cast isn't safe. Even if the superclass is doing a type check you're only guaranteed to have a ClusterProperties object. Unfortunately you have the general problem of non-symmetric equals since you have a subclass situation and the it may return true when comparing two objects of different types, but false when the comparison order is reversed. See item 10 from Effective Java for more info. I don't see a good way around this given other constraints, but the cast should be type checked first at least.

Reply:
the super class's type check "getClass() != obj.getClass()" actually will fail if we are comparing a ClusterStoreProperties with a ClusterProperties (checked by stepping thru the code). It will see "ClusterStoreProperties" != "ClusterProperties". 
So only when the class names are exactly the same, (comparing a ClusterProperties with a ClusterProperties, or a ClusterStoreProperties with a ClusterStoreProperties), the super class type check will pass. This means at line 178 in this class, the class type is guaranteed ClusterStoreProperties.

Reply:
Okay, thanks for confirming. In general it's not safe to assume the behavior of a super class will remain unchanged - doing the type check here for safety and clarity is still a good idea.

Reply:
makes sense. fixed. thanks.